src/julia_gc.c

```diff
@@ -0,0 +1,824 @@
+            if (R(t)->prio > t->prio) {
+                treap_rot_left(treap);
+            }
+        }
+    }
+}
+
+static void treap_delete_node(treap_t ** treap)
+{
+    for (;;) {
+        treap_t * t = *treap;
+        if (L(t) == NULL) {
+            *treap = R(t);
+            free_treap(t);
+            break;
+        }
+        else if (R(t) == NULL) {
+            *treap = L(t);
+            free_treap(t);
+            break;
+        }
+        else {
+            if (L(t)->prio > R(t)->prio) {
+                treap_rot_right(treap);
+                treap = &R(*treap);
+            }
+            else {
+                treap_rot_left(treap);
+                treap = &L(*treap);
+            }
+        }
+    }
+}
+
+static int treap_delete(treap_t ** treap, void * addr)
+{
+    while (*treap != NULL) {
+        int c = cmp_ptr(addr, (*treap)->addr);
+        if (c == 0) {
+            treap_delete_node(treap);
+            return 1;
+        }
+        else if (c < 0) {
+            treap = &L(*treap);
+        }
+        else {
+            treap = &R(*treap);
+        }
+    }
+    return 0;
+}
+
+static uint64_t xorshift_rng_state = 1;
+
+static uint64_t xorshift_rng(void)
+{
+    uint64_t x = xorshift_rng_state;
+    x = x ^ (x >> 12);
+    x = x ^ (x << 25);
+    x = x ^ (x >> 27);
+    xorshift_rng_state = x;
+    return x * (uint64_t)0x2545F4914F6CDD1DUL;
+}
+
+
+static treap_t * bigvals;
+
+void alloc_bigval(void * addr, size_t size)
+{
+    treap_t * node = alloc_treap();
+    node->addr = addr;
+    node->size = size;
+    node->prio = xorshift_rng();
+    treap_insert(&bigvals, node);
+}
+
+void free_bigval(void * p)
+{
+    if (p) {
+        treap_delete(&bigvals, p);
+    }
+}
+
+static jl_module_t *   Module;
+static jl_datatype_t * datatype_mptr;
+static jl_datatype_t * datatype_bag;
+static jl_datatype_t * datatype_largebag;
+static Bag *           StackBottomBags;
+static UInt            StackAlignBags;
+static jl_ptls_t       JuliaTLS, SaveTLS;
+static size_t          max_pool_obj_size;
+static size_t          bigval_startoffset;
+static UInt            YoungRef;
+
+
+#ifndef NR_GLOBAL_BAGS
+#define NR_GLOBAL_BAGS 20000L
+#endif
+
+static Bag *        GlobalAddr[NR_GLOBAL_BAGS];
+static const Char * GlobalCookie[NR_GLOBAL_BAGS];
+static Int          GlobalCount;
+
+
+/****************************************************************************
+**
+*F  AllocateBagMemory( <type>, <size> )
+**
+**  Allocate memory for a new bag.
+**/
+static void * AllocateBagMemory(UInt type, UInt size)
+{
+    // HOOK: return `size` bytes memory of TNUM `type`.
+    void * result;
+    if (size <= max_pool_obj_size) {
+        result = (void *)jl_gc_alloc_typed(JuliaTLS, size, datatype_bag);
+    }
+    else {
+        result = (void *)jl_gc_alloc_typed(JuliaTLS, size, datatype_largebag);
+    }
+    memset(result, 0, size);
+    if (TabFreeFuncBags[type])
+        jl_gc_set_needs_foreign_finalizer((jl_value_t *)result);
```

Comment:
I just realized that this is not quite enough: we also need this in `RetypeBag`. And then we may also want to be able to undo this, i.e. add a second argument `int needed` to `jl_gc_set_needs_foreign_finalizer`, and replace the above by
```C
  jl_gc_set_needs_foreign_finalizer((jl_value_t *)result), TabFreeFuncBags[type] != 0);
```
and add the same to `RetypeBag`. Thoughts?

Reply:
Actually, I take that back: I don't think Julia allows you to remove a finalizer. But there also is no need for that: Changing an object from one that does need a finalizer to one that does not should be very rare (if it happens at all), and we can just change `JFinalizer` to not do anything in that case.